### PR TITLE
Connect integrations

### PR DESCRIPTION
**Pull request by AWinterman**

### CODE REVIEW

File: connect-integrations.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+> A tributary is a stream or river that flows into a larger
+> stream or main stem (or parent) river or a lake. A tributary does not flow
+> directly into a sea or ocean. Tributaries and the main stem river drain the
+> surrounding drainage basin of its surface water and groundwater, leading the
+> water out into an ocean.
+
+Urban Airship Connect is a streaming HTTP API which provides near-real-time
+mobile event data. The goal of the project is to give customers access to 
+the data we collect on their behalf. However, many customers don't want to
```
Comment (nbanerje) [inline connect-integrations.md]: We might want to include something like, ... give customers user level data that we collect on their behalf and enrich with additional information.

File: connect-integrations.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+spend the resources building a bespoke pipeline from our API to their analytics 
+engine. Enter Tributary, which takes the stream of Connect mobile data into 
+the ocean of third-party analytics and data storage tools.
```
Comment (nbanerje) [inline connect-integrations.md]: Might want to say 3rd party analytics, data storage, and marketing tools.

File: connect-integrations.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+Tributary is a distributed system that can scale arbitrarily with customer
+demand. It's functionality is split across hardware nodes on our internal VPN, 
+and AWS hosted services. It relies particularly heavily on Amazon Web Services
+Elastic Cloud Compute Service (AWS ECS), which essentially ensures that the docker
+containers you'd like to be running are. This doesn't guarantee that your
+processes are running: ECS can fail to find a node that has enough resources to
+fulfill your specification, or configuration errors might prevent the process
+inside the docker container from ever doing any work, but it does mean that you
+can just shutdown the process in case of exception, and be confident that it
+will get started back up again.
```
Comment (nbanerje) [inline connect-integrations.md:28]: Are there particular annoying problems that are avoided by this or another strategy that this is superior to?

File: connect-integrations.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+was a problem, we'd get literally hundreds of pages all at once, which really
+isn't helpful. If a problem was due to misconfiguration or programming error,
+it would alert literally hundreds of times.
+
+The problem boils down to the fact that each service has no concept of what's going
+on in the rest of the system, let alone what has happened in the system in the past.
+It knows that it encountered an exception, but has no way of knowing whether the 
+equivalent exception has already been encountered by another integration, or even 
+whether this is the umpteenth time the integration in question has encountered the 
+exception Moreover, we're now in the business of classifying exceptions-- we need to
```
Comment (nbanerje) [inline connect-integrations.md:131]: I think you need a . btw exception and Moreover

File: connect-integrations.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+The controller service scans the dynamo table, and keeps counts of how many 
+integrations have encountered exceptions, how many are lagged, and how many
+are starting up very frequently. It sends these counts to a 
+prometheus AlertManager, which notifies us if a sufficient proportion of them 
+are in alerting territory. 
+
+The goal is to never page unless the problem is real, but also to make sure we
+know when something is wrong. I think this balance does that-- intermittent
+issues are surfaced by sentry, on a timeline that is convenient for the
+developer. Major issues are surfaced immediately by our system level alerting.
```
Comment (nbanerje) [inline connect-integrations.md:207]: I think a strong conclusion could be about the ease of writing configurations, but the biggest amount of work for this type of system is the monitoring and alerting.

**Review by nbanerje — APPROVED** [review]

**Comment by nbanerje** [comment]
:+1:

**Comment by AWinterman** [comment]
Closing this, it's been years.

**Pull request closed without merging**